.github/workflows/samples-tools/cmd/affected/main.go

```diff
@@ -0,0 +1,84 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"slices"
+
+	"samples-tools/pkg/utils"
+)
+
+func main() {
+	usage := "usage: affected path/to/config.json [head-commit] [main-commit]"
+	configPath := utils.ArgRequired(1, usage)
+	headCommit := utils.ArgWithDefault(2, "HEAD")
+	mainCommit := utils.ArgWithDefault(3, "origin/main")
+
+	config, err := utils.LoadConfig(configPath)
+	if err != nil {
+		panic(err)
+	}
+
+	diffs, err := utils.Diffs(headCommit, mainCommit)
+	if err != nil {
+		panic(err)
+	}
+
+	matrix, err := affected(config, diffs)
+	if err != nil {
+		panic(err)
+	}
+
+	matrixJson, err := json.Marshal(matrix)
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Println(string(matrixJson))
+}
+
+func affected(config utils.Config, diffs []string) ([]string, error) {
+	uniquePackages := make(map[string]bool)
+	for _, diff := range diffs {
+		if !config.Matches(diff) {
+			continue
+		}
+		pkg := config.FindPackage(diff)
+		uniquePackages[pkg] = true
+	}
+
+	changed := make([]string, 0, len(uniquePackages))
+	for pkg := range uniquePackages {
+		changed = append(changed, pkg)
+	}
+
+	if !slices.Contains(changed, ".") {
+		return changed, nil
+	}
+
+	var packages []string
+	err := filepath.WalkDir(".",
```

Comment:
Thanks! Sounds good. I changed it to use `io/fs.WalkDir` and additionally surfaced the root directory as a parameter. That way we could eventually add tests and specify our own path rather than always using the current directory.